### PR TITLE
Move DeltaManager.active handling out of DeltaManager

### PR DESCRIPTION
**Pull request by vladsud**
While glancing at why some code many not work, I've noticed that code around DeltaManager.active implementation uses wrong terms.
For example,  this._deltaManager.updateQuorumJoin() call in Container from "connected" handler, where "in quorum" is not valid way to describe current state - it relies on us being in "write" mode to be valid (while this code executes for both "read" & "write"). The end-to-end code is correct, but it's easy to conclude otherwise glancing at each layer separatly. Also DeltaManager ideally should have nothing to do with quorum.

So I prototyped getting this info out of DeltaManager and having active() property being implemented in Container, where actual knowledge about what is active is. It's a bit ugly, but I think it moves logic into right place.

For some reason my attempt with Omit<> did not yield what I wanted - tsc would still complain about unrelated functions (on, off), if I do not add active property to DeltaManager. Any hints on why it might be happening would be appreciated.

Thoughts?

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -1430,7 +1426,20 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             this.emit("readonly", readonly);
         });
 
-        return deltaManager;
+        // Add "active" property.
+        // The meaning of it is Container specific, as we allow container runtime to send ops
+        // only when we are connected as "write" and are up to date, i.e. observed our own join op.
+        // The later info does not exists at DeltaManager layer, it's Container specific logic.
+        Object.defineProperties(deltaManager, {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:1433]: i think i'd rather just pass in a callback, rather than do this

Comment (vladsud) [reply packages/loader/container-loader/src/container.ts:1433]: I thought about callback, but my ideal solution would be not not even expose this concept in DeltaManager.
I guess there is no good choice here :(

**Comment by agarwal-navin** [comment]
I am trying to understand what is the significance of "active" here. What does being active mean for a client? Is it to let apps know that we are not in "write" mode so that they can block users from writing anything?

I am currently discussing an issue with Bohemia where they using the active signal to decide whether to create a collection component or not. So, trying to understand what the intention for active property is.

**Comment by vladsud** [comment]
"Active" signifies the fact that IRuntime can send ops. If active === false, then runtime can not send any ops.
Ideally this concept should not be at Loader level, as it can be easily computed in ContainerRuntime layer as simply being "connected" with "write" connection. 

My long term view on that is that we should "compose" IDeltaManager object as we move through layers. I.e. it may take even different forms as it gets exposed to host, container runtime, data store runtime. But there is always a single object at a given layer that exposes all the info about "connection" that this layer need to have access (and I think things like audience, quorum are on this interface), but it's not same thing that is exposed by DeltaManager.
And while Container layer is the one who controls the logic when IRuntime receives "connected" event (and switched clientId), that is the only piece of info expose to IRuntime and it can define what is "active" and expose on some type like IDeltaManager.

I actually played here with Object.create as a way to compose / build adapters, but it does not work how I expected. So I'm not sure what's the best way to compose / modify objects other than manually build pimpl wrappers (that would suck).

**Comment by agarwal-navin** [comment]
This maybe a dumb question but the runtime can send ops right. It's just that we would get nack'd, reconnect in write mode and resend this op. Do you mean that this is just how the implementation currently is and ideally you cannot send op until you explicitly change mode to write?

**Comment by vladsud** [comment]
Runtime can't send ops until active === true. It's not just about nacks. If you are connected as "write", but active === false, that means we did not caught up, and thus we have not had a chance to observe all of our own acks, and thus sending an op would potentially result in same op being sent twice.

Notice that delta manager has nothing to do with it - it has no business of knowing when it's save to send ops (it has also nothing to do with quorum today). Tough this is half true, as I can point you the code that has to know at least small bit about it (we reconnect as "write" on lost connection if DeltaManager detected that there were some ops that never acked in previous connection, not doing so will result in same data loss around double resend).

**Review by anthony-murphy — APPROVED** [review]

**Comment by vladsud** [comment]
@anthony-murphy , please take a look - it changed a bit after your sign off. The issue I've hit - automatic noop sending is unfortunately depends of active state. It logically belongs to same layer where protocol handling is (Container, not DeltaManager) as we automatically send no-ops as an ack to proposals if there was no rejection, and we do it only in active mode.

One non-obvious side-effect of current design (unchanged) is that client can be in quorum, but not active, and thus automatically accepts all proposal, but does it only when it becomes active. This is a bit weird (as it either should auto-accept right away, or design should allow refusals at a time when it catches up). The worst part of current design - it can hold collab window for very long time... Opened https://github.com/microsoft/FluidFramework/issues/5150 to track it.

I've also removed submit from IDeltaSender, as it does not belong there - users of DeltaManager should not be able to submit ops that way, they always should go through ContainerRuntime -> Container -> DeltaManager.

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -1342,12 +1396,22 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         return client;
     }
 
+    /**
+     * Returns true if connection is active, i.e. it's "write" connection and
+     * container runtime was notified about this connection (i.e. we are up-to-date and could send ops)
```
Comment (curtisman) [inline packages/loader/container-loader/src/container.ts:1401]: might want to mention that you are in the quorum or you have received the Join op, which is the condition for be `Connected` in write mode.

**Review by curtisman — APPROVED** [review]
:shipit:

File: packages/loader/container-loader/src/container.ts [hunk 3]
```diff
@@ -182,6 +182,60 @@ export async function waitContainerToCatchUp(container: Container) {
...
+        // operation. This allows the server to know our true reference sequence number and be able to
+        // correctly update the minimum sequence number (MSN).
+        if (this.updateSequenceNumberTimer === undefined) {
+            // Clear an update in 2 s
+            this.updateSequenceNumberTimer = setTimeout(() => {
+                this.updateSequenceNumberTimer = undefined;
+                if (this.activeConnection()) {
+                    this.submit(MessageType.NoOp, null);
+                }
+            }, 2000);
```
Comment (curtisman) [inline packages/loader/container-loader/src/container.ts:227]: Might want to make this a constant variable so we can tune it?

**Pull request merged**